### DISCUSSION TITLE
No performance improvement in OpenSSL3.2.1 over OpenSSL3.0.13

### DISCUSSION THREAD

**Discussion by bjayanax** [openssl/openssl#23556] [Q&A]
We ran few ciphers in OpenSSL3.2.1 & OpenSSL3.0.13 and not observed any performance improvement in OpenSSL3.2.1. Adding cipher details and performance values for reference.

Cipher suite: AES128-SHA
TLS version: TLSv1.2
<img width="241" alt="image" src="https://github.com/openssl/openssl/assets/96256276/9c35bf15-1d96-4f24-b63a-44232c379ac7">

Cipher suite: ECDHE-RSA-AES128-SHA
TLS version: TLSv1.2
<img width="239" alt="image" src="https://github.com/openssl/openssl/assets/96256276/e4d70fc1-2937-4e4d-b0c9-b41bb40364b5">

Cipher suite: ECDHE-ECDSA-AES128-GCM-SHA256
curve x25519 & cert P256
TLS version: TLSv1.2
<img width="239" alt="image" src="https://github.com/openssl/openssl/assets/96256276/10d29398-8a47-4e03-a08a-9df228487553">

Cipher suite: ECDHE-ECDSA-AES128-GCM-SHA256
curve & cert P256
TLS version: TLSv1.2
<img width="240" alt="image" src="https://github.com/openssl/openssl/assets/96256276/fdb1f5e7-affc-44f1-b499-617c9b3fdeb5">

As mentioned earlier ([OpenSSL#21833](https://github.com/openssl/openssl/issues/21833) comments) not seen any performance improvement in OpenSSL3.2 . 

<img width="664" alt="image" src="https://github.com/openssl/openssl/assets/96256276/34fbd3b2-4f78-461f-869f-3086587a1bbb">

**Comment by paulidale** [comment]
Why would you expect raw crypto primitives to be faster?
Do a handshakes per second test in a multiprocess environment and you'll see a significant change.

**Reply by bjayanax** [reply]
We have ran nginx performance upto 64 process and not seen any performance improvement in OpenSSL3.2. 

Cipher suite: AES128-SHA
TLS version: TLSv1.2
<img width="301" alt="image" src="https://github.com/openssl/openssl/assets/96256276/34d609ce-d123-4ca1-b441-b67ef45b6fac">

Cipher suite: ECDHE-RSA-AES128-SHA
TLS version: TLSv1.2
<img width="301" alt="image" src="https://github.com/openssl/openssl/assets/96256276/b3b1e612-55c3-47f2-8d81-f5acd499df87">

Cipher suite: ECDHE-ECDSA-AES128-GCM-SHA256
curve x25519 & cert P256
TLS version: TLSv1.2
<img width="299" alt="image" src="https://github.com/openssl/openssl/assets/96256276/c0901298-cb6d-4802-b485-f4d67ff6f54e">


Cipher suite: ECDHE-ECDSA-AES128-GCM-SHA256
curve & cert P256
TLS version: TLSv1.2
<img width="300" alt="image" src="https://github.com/openssl/openssl/assets/96256276/1155d899-598b-41c4-9a3f-952b64fafc75">

**Reply by t8m** [reply]
What are the numbers measuring? Is that handshake latency, throughput or something else?  Are the worker processes real separate processes or a single process with that number of concurrent threads?

**Reply by bjayanax** [reply]
We ran nginx Handshake and got above numbers. Mentioned values are CPS(Connections per Seconds). Worker processes are not threads those are separate process.

**Reply by t8m** [reply]
Hmm... that might explain some things because many of the changes in 3.1 and 3.2 improved just concurrent threads handling. But I still assume all the worker processes do multiple connections not just one and thus some of the caching added in 3.2 should improve the performance. Is that openssl-3.2.1 built with no-quic or without it? There is some (small) overhead caused by the added QUIC support.

**Reply by bjayanax** [reply]
During OpenSSL built time not used any additional flags. 

configuration command used: 
./config --prefix=/root/user_name/openssl_install/

**Comment by nhorman** [comment]
can you describe the nginx handshake command (I don't see it documented anywhere in the nginx docs), and your test environment?  specifically:
1) How many cpus on the server and clients
2) What type of network setup connects these systems

**Reply by bjayanax** [reply]
Commands :
	Server command: ./nginx
	Client command: ./openssl s_time -connect server_IP:443 -new -cipher AES128-SHA -time 50
	Note: by using our internal script pumping traffic (running client) 2000 times
	
	For this test, we used a server with four clients. All are connected physically.
	CPU info:
		Server have 240 CPU's
		Client 1 & 2 have 144 CPU's
		Client 3 & 4 have 160 CPU's.
